components/brave_sync/qr_code_validator.cc

```diff
@@ -0,0 +1,106 @@
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/brave_sync/qr_code_validator.h"
+
+#include <string>
+#include <vector>
+
+#include "base/logging.h"
+#include "base/strings/string_number_conversions.h"
+#include "base/strings/string_util.h"
+#include "base/time/time.h"
+#include "brave/components/brave_sync/crypto/crypto.h"
+#include "brave/components/brave_sync/qr_code_data.h"
+
+namespace brave_sync {
+
+namespace {
+
+static constexpr base::TimeDelta kIntervalForValidForTooLong =
+    base::Minutes(60);
+static constexpr size_t kSeedBytesCount = 32u;
+static constexpr char kQRv1SunsetDate[] = "Sat, 1 Jan 2022 00:00:00 GMT";
+
+}  // namespace
+
+base::Time QrCodeDataValidator::qr_v1_sunset_day_;
+
+base::Time QrCodeDataValidator::GetQRv1SunsetDay() {
+  if (qr_v1_sunset_day_.is_null()) {
+    bool convert_result =
+        base::Time::FromUTCString(kQRv1SunsetDate, &qr_v1_sunset_day_);
+    CHECK(convert_result);
+  }
+
+  CHECK(!qr_v1_sunset_day_.is_null());
+
+  return qr_v1_sunset_day_;
+}
+
+QrCodeDataValidationResult QrCodeDataValidator::ValidateQrDataJson(
+    const std::string& qr_data_string) {
+  auto now = base::Time::Now();
+  auto qr_code_data = QrCodeData::FromJson(qr_data_string);
+
+  if (!qr_code_data) {
+    if (IsValidSeedHex(qr_data_string)) {
+      if (now < GetQRv1SunsetDay()) {
+        return QrCodeDataValidationResult::kValid;
+      } else {
+        return QrCodeDataValidationResult::kVersionDeprecated;
+      }
+    }
+    return QrCodeDataValidationResult::kNotWellFormed;
+  }
+
+  if (qr_code_data->sync_code_hex.empty()) {
+    return QrCodeDataValidationResult::kNotWellFormed;
+  }
+
+  if (qr_code_data->version < QrCodeData::kCurrentQrCodeDataVersion) {
+    return QrCodeDataValidationResult::kVersionDeprecated;
+  } else if (qr_code_data->version > QrCodeData::kCurrentQrCodeDataVersion) {
+    return QrCodeDataValidationResult::kVersionNotRecognized;
+  }
+
+  if (!IsValidSeedHex(qr_code_data->sync_code_hex)) {
+    return QrCodeDataValidationResult::kNotWellFormed;
+  }
+
+  if (now > qr_code_data->not_after) {
+    return QrCodeDataValidationResult::kExpired;
+  }
+
+  if (qr_code_data->not_after - now > kIntervalForValidForTooLong) {
+    return QrCodeDataValidationResult::kValidForTooLong;
+  }
+
+  return QrCodeDataValidationResult::kValid;
+}
+
+bool QrCodeDataValidator::IsValidSeedHex(const std::string& seed_hex) {
+  std::string seed_hex_trimmed;
+  base::TrimString(seed_hex, " \n\t", &seed_hex_trimmed);
+
+  std::vector<uint8_t> bytes;
+  if (!base::HexStringToBytes(seed_hex_trimmed, &bytes)) {
+    return false;
+  }
+
+  if (bytes.size() != kSeedBytesCount) {
+    return false;
+  }
+
+  std::string sync_code_words =
+      brave_sync::crypto::PassphraseFromBytes32(bytes);
+  if (sync_code_words.empty()) {
```

Comment:
Not essential, but maybe we could also check that the number of words is 24 here.

Reply:
Thanks for this comment

With fresh eyes for me it seems that `brave_sync::crypto::PassphraseFromBytes32` must succeed for any (at least non-null) data.  So the current check 
```
if (sync_code_words.empty()) {
  return false;
}
```
for sure seems useless.

About compare for the the words count I think that  for  `kSeedBytesCount` length data the words count  also is some defined constant in any way.

There are two ways to change
1. remove `if (sync_code_words.empty()) { ... ` - makes code simpler, removes `#include` and `deps += :crypto`
2. keep `if (sync_code_words.empty()) { ... ` and add `if (words_count != kExpectedWordsCount) - worth to use if it is normal situation when number of words may vary.

I would go with pt1. 

@fmarier , @darkdh , what would you say?

Reply:
Since this is a very critical portion of the code (it defines was encryption key ends up being used by clients), I would suggest we err on the side of being extra-cautious. So I wouldn't remove the check because even though it may be redundant right now, it could catch a very serious bug in the future.

So I would go with the word count check (option 2). When we implement the 25-word variant of the safety words (after Phase 1), then this check will be even more useful.

Reply:
sync words (mnemonic) length depends on the entropy size, 32 bytes-> 24 words and if it is not 24 words that means the library we use has bug so it doesn't harm to do extra check. Probably add `NOTREACHED()` so we can easily catch it on debug build when that happens

Reply:
thanks, I have updated at 
https://github.com/brave/brave-core/pull/11199/commits/fa9cdb4ed6af96730f7c2b6c2a4a7b59ec22bea5#diff-5cc9f02f78171f50ac39e806f4cd431906d5581cc97e812022206c55381c3f2cR110